### PR TITLE
win_chocolatey_source: add new module to manage Chocolatey sources

### PR DESCRIPTION
**Pull request by jborean93**
##### SUMMARY
Added new module to manage sources within Chocolatey.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
win_chocolatey_source

##### ANSIBLE VERSION
```
devel
```

### CODE REVIEW

File: lib/ansible/modules/windows/win_chocolatey_source.ps1 [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+        $choco_app,
+        $name,
+        $attribute,
+        $value
+    )
+    # for private attributes like password/certificatePassword, the only way
+    # for us to preserve the existing value when changing another attirbute is
+    # to store it manually by editing the XML
+    $choco_config_path = "$(Split-Path -Path (Split-Path -Path $choco_app.Path))\config\chocolatey.config"
+    if (-not (Test-Path -Path $choco_config_path)) {
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_chocolatey_source.ps1]: When I see `-Path` in Powershell nowadays, I always wonder *Shouldn't this be `-LiteralPath` instead ?*

File: lib/ansible/modules/windows/win_chocolatey_source.ps1 [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+        $allow_self_service,
+        $admin_only
+    )
+    # build the base arguments
+    $arguments = [System.Collections.ArrayList]@($choco_app.Path,
+        "source", "add", "--name", $name, "--source", $source
+    )
+
+    # add optional arguments from user input
+    if ($null -ne $source_username) {
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_chocolatey_source.ps1:133]: So turning this around here is not needed, as $source_username cannot be a collection.
I think that's why we disabled that rule in pslint, so we didn't *have* to comply with something that is futile.

File: lib/ansible/modules/windows/win_chocolatey_source.ps1 [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+    }
+
+    if ($check_mode) {
+        $arguments.Add("--what-if") > $null
+    }
+
+    $command = Argv-ToString -arguments $arguments
+    $res = Run-Command -command $command
+    if ($res.rc -ne 0) {
+        Fail-Json -obj $result -message "Failed to add Chocolatey source '$name': $($_.stderr)"
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_chocolatey_source.ps1]: I was surprised to see **$_** in this context, should that not be **$rc.stderr** instead ?

**Review by dagwieers — APPROVED** [review]
A few remarks, nothing much else to add.

File: lib/ansible/modules/windows/win_chocolatey_source.ps1 (hunk 2 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_chocolatey_source.ps1:133]: You are right it isn't needed here but I've started to get into the habit of doing this everywhere. I don't want to force other people to comply but I have started to do it myself.

File: lib/ansible/modules/windows/win_chocolatey_source.ps1 (hunk 3 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_chocolatey_source.ps1]: You are right, joys of copying and pasting liberally.

File: lib/ansible/modules/windows/win_chocolatey_source.ps1 (hunk 1 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_chocolatey_source.ps1]: Wouldn't hurt, probably isn't necessary as the path source is from `Get-Command` but will change it anyway.

File: lib/ansible/modules/windows/win_chocolatey_source.ps1 [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+
+Function Get-ChocolateySources {
+    param($choco_app)
+
+    $choco_config_path = "$(Split-Path -Path (Split-Path -Path $choco_app.Path))\config\chocolatey.config"
+    if (-not (Test-Path -LiteralPath $choco_config_path)) {
+        Fail-Json -obj $result -message "Expecting Chocolatey config file to exist at '$choco_config_path'"
+    }
+
+    # would prefer to enumerate the existing sources with an actual API but the
```
Comment (gep13) [inline lib/ansible/modules/windows/win_chocolatey_source.ps1:48]: @jborean93 could I ask that you raise an issue about this on chocolatey/choco?  There might already be an issue about this, so if you could check that would be great.

**Comment by gep13** [comment]
@jborean93 thanks for doing the work to get this added! 😄 

If/when this is added to Ansible, a PR to update to this page:

https://chocolatey.org/docs/features-infrastructure-automation#chocolatey-integration-implementation-with-common-configuration-managers

Would be greatly appreciated!

File: lib/ansible/modules/windows/win_chocolatey_source.ps1 (hunk 2 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_chocolatey_source.ps1:133]: I had hoped it wouldn't have come this far, but I don't disagree. At least it's easier than second guessing or leaving it up to a static analyzer (which are still not perfect wrt. this).

File: lib/ansible/modules/windows/win_chocolatey_source.ps1 (hunk 4 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_chocolatey_source.ps1:48]: I had a brief search and couldn't fine one, have raised the issue here https://github.com/chocolatey/choco/issues/1614.

**Comment by dagwieers** [comment]
@gep13 @jborean93 It could be useful to extend the following page: https://docs.ansible.com/ansible/2.5/user_guide/windows_usage.html#installing-software

And maybe have a dedicated Chocolatey page so that it can be referenced from the Chocolatey documentation instead of (only) the **win_chocolatey** manual page.

**Comment by gep13** [comment]
@dagwieers that seems like a great idea to me!  I will be honest and say that I have never used Ansible in anger, so I don't know if I would be the right person to do this work, but perhaps @jborean93 would be in a position to help.  I could certainly help with any questions regarding Chocolatey specifically.

**Comment by gep13** [comment]
@dagwieers I can certainly make the change to the link on the chocolatey.org site happen.

**Comment by jborean93** [comment]
I was looking at either expanding our docs or the Chocolatey docs once getting some of these new modules and roles in. I've been slowly working through a lot features/functions in Chocolatey to get a better handle of it all and improving our use of it in Ansible.

**Comment by dagwieers** [comment]
cc @ferventcoder

**Comment by dagwieers** [comment]
@gep13 @ferventcoder If there's an interest, we could add you (or other Chocolatey devs) to @ansibot so that you would be notified if any **win_chocolatey**-related bug or PR is opened or discussed (on average less than 4 a month).

**Comment by gep13** [comment]
@dagwieers I would have no objections to being added to that.

**Review by nitzmahone — APPROVED** [review]
I'm a little surprised their `--what-if` mode works the way it does there (and it lies about whether it would have made a change, but we don't currently care)... So, LGTM!

**Pull request merged**